src/app/clusters/network-commissioning/network-commissioning.cpp

```diff
@@ -371,18 +449,122 @@ void Instance::HandleAddOrUpdateWiFiNetwork(HandlerContext & ctx, const Commands
+                }
+                networkIndex++;
+                return Loop::Continue;
+            });
+            if (!clientIdentityNetworkIndex.HasValue())
+            {
+                ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Protocols::InteractionModel::Status::InvalidCommand,
+                                              "clientIdentifier");
+                return;
+            }
+        }
+
+        // Allocate a buffer to hold the client identity, and leave enough room to append the possession nonce if needed.
+        chip::Platform::ScopedMemoryBuffer<uint8_t> identityBuffer;
+        size_t identityBufferSize = kMaxCHIPCompactNetworkIdentityLength + (provePossession ? kPossessionNonceSize : 0);
+        VerifyOrExit(identityBuffer.Alloc(identityBufferSize).Get() != nullptr, /**/);
+
+        // Add/Update the network at the driver level
+        MutableByteSpan clientIdentity(identityBuffer.Get(), kMaxCHIPCompactNetworkIdentityLength);
+        Optional<P256ECDSASignature> possessionSignature;
+        DebugTextStorage debugTextBuffer;
+        MutableCharSpan debugText(debugTextBuffer);
+        uint8_t networkIndex;
+        auto && result = driver->AddOrUpdateNetworkWithPDC(req.ssid, networkIdentity, clientIdentityNetworkIndex, debugText,
+                                                           clientIdentity, networkIndex);
+
+        // Map CHIP_ERROR (other than CHIP_NO_ERROR) to an IM Failure status rather than a NetworkConfigResponse
+        VerifyOrExit(result.Is<Status>() || (result.Is<CHIP_ERROR>() && result.Get<CHIP_ERROR>() == CHIP_NO_ERROR), /**/);
+
+        Commands::NetworkConfigResponse::Type response;
+        response.networkingStatus = (result.Is<Status>()) ? result.Get<Status>() : Status::kSuccess;
+        FillDebugTextAndNetworkIndex(response, debugText, networkIndex);
+
+        if (response.networkingStatus == Status::kSuccess)
+        {
+            response.clientIdentity.SetValue(clientIdentity);
+
+            if (provePossession)
+            {
+                // PossessionSignature TBS message = (NetworkClientIdentity || PossessionNonce)
+                memcpy(clientIdentity.end(), req.possessionNonce.Value().data(), kPossessionNonceSize);
+                ByteSpan proofTbsMessage(clientIdentity.data(), clientIdentity.size() + kPossessionNonceSize);
+                SuccessOrExit(driver->SignWithClientIdentity(networkIndex, proofTbsMessage, possessionSignature.Emplace()));
+                response.possessionSignature.SetValue(possessionSignature.Value().Span());
+            }
+
+            UpdateBreadcrumb(req.breadcrumb);
+        }
+
+        SuccessOrExit(ctx.mCommandHandler.AddResponseData(ctx.mRequestPath, response));
+        status = Protocols::InteractionModel::Status::Success;
+    }
+
+exit:
+    if (status != Protocols::InteractionModel::Status::Success)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Protocols::InteractionModel::Status::Failure);
```

Comment:
Shouldn't we just add the actual status here?

Reply:
Tweaked this now to track a CHIP_ERROR within the method to follow the `SuccessOrExit(err = ...)` convention.